### PR TITLE
Read certificates from storage in batches of 100

### PR DESCRIPTION
**Pull request by deuszx**
## Motivation

A malicious actor could send a `download_certificates(hashes)`  – with `hashes` being a very long vector – and cause a memory leak (or even crash an app) b/c we are not limiting the number of certificates we read from storage at once.

## Proposal

Read certificates from the underlying storage in batches of 100 at once.

## Release Plan

- Nothing to do / These changes follow the usual release cycle.

## Links
Closes https://github.com/linera-io/linera-protocol/issues/2658

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-service/src/proxy/grpc.rs [hunk 1]
```diff
@@ -491,17 +491,19 @@ where
...
+                .0
+                .storage
+                .read_certificates(batch.to_vec())
+                .await
+                .map_err(|err| Status::from_error(Box::new(err)))?
+            {
+                if grpc_message_limiter.fits::<Certificate>(certificate.clone())? {
+                    certificates.push(certificate);
+                } else {
+                    break;
```
Comment (afck) [inline linera-service/src/proxy/grpc.rs:505]: We should break out of the _outer_ loop here.

Comment (deuszx) [reply linera-service/src/proxy/grpc.rs:505]: Ah, good point.

Comment (deuszx) [reply linera-service/src/proxy/grpc.rs]: Done (pushed with force).

**Review by afck — APPROVED** [review]

File: linera-service/src/proxy/grpc.rs [hunk 2]
```diff
@@ -491,17 +491,19 @@ where
...
+                .await
+                .map_err(|err| Status::from_error(Box::new(err)))?
+            {
+                if grpc_message_limiter.fits::<Certificate>(certificate.clone())? {
+                    certificates.push(certificate);
+                } else {
+                    break 'outer;
+                }
             }
         }
```
Comment (andresilva91) [inline linera-service/src/proxy/grpc.rs:509]: <i class='graphite__hidden'>[Re: line +511]<br/><br/></i>

Might be worth to log the number of certificates we're returning from here in a prometheus gauge or histogram, so we know how many certificates we're fitting in the limit<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2663?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: linera-service/src/proxy/grpc.rs [hunk 3]
```diff
@@ -477,7 +477,7 @@ where
         &self,
         request: Request<CertificatesBatchRequest>,
     ) -> Result<Response<CertificatesBatchResponse>, Status> {
-        let hashes = request
+        let hashes: Vec<linera_base::crypto::CryptoHash> = request
```
Comment (ma2bd) [inline linera-service/src/proxy/grpc.rs:480]: @deuszx This annotation was not needed

File: linera-service/src/proxy/grpc.rs [hunk 4]
```diff
@@ -491,17 +491,19 @@ where
...
-            .storage
-            .read_certificates(hashes)
-            .await
-            .map_err(|err| Status::from_error(Box::new(err)))?
-        {
-            if grpc_message_limiter.fits::<Certificate>(certificate.clone())? {
-                certificates.push(certificate);
-            } else {
-                break;
+        'outer: for batch in hashes.chunks(100) {
```
Comment (ma2bd) [inline linera-service/src/proxy/grpc.rs:494]: Can we define a constant instead of using `100`?

Comment (ma2bd) [reply linera-service/src/proxy/grpc.rs:494]: (or better a proper config option)

File: linera-service/src/proxy/grpc.rs [hunk 5]
```diff
@@ -491,17 +491,19 @@ where
```
Comment (ma2bd) [inline linera-service/src/proxy/grpc.rs:491]: I feel that we're going to need a config option for this, soon or later.

Comment (deuszx) [reply linera-service/src/proxy/grpc.rs:491]: I can add it but I don't think it'll be necessary: with 16MB of gRPC message size limit a requester validator will never receive more than a couple of certificates.

Comment (ma2bd) [reply linera-service/src/proxy/grpc.rs:491]: Let's check the data

**Pull request merged**